rasa/shared/core/domain.py

```diff
@@ -270,58 +273,172 @@ def merge(self, domain: Optional["Domain"], override: bool = False) -> "Domain":
         domain_dict = domain.as_dict()
         combined = self.as_dict()
 
-        def merge_dicts(
-            tempDict1: Dict[Text, Any],
-            tempDict2: Dict[Text, Any],
-            override_existing_values: bool = False,
-        ) -> Dict[Text, Any]:
+        if override:
+            config = domain_dict["config"]
+            for key, val in config.items():
+                combined["config"][key] = val
 
-            if override_existing_values:
-                merge_dicts, b = tempDict1.copy(), tempDict2.copy()
+        if override or self.session_config == SessionConfig.default():
+            combined[SESSION_CONFIG_KEY] = domain_dict[SESSION_CONFIG_KEY]
 
-            else:
-                merge_dicts, b = tempDict2.copy(), tempDict1.copy()
+        if combined[KEY_INTENTS] or domain_dict[KEY_INTENTS]:
+            combined[KEY_INTENTS] = self.merge_lists_of_dicts(
+                combined[KEY_INTENTS], domain_dict[KEY_INTENTS], override
+            )
+        # remove existing forms from new actions
+        for form in combined[KEY_FORMS]:
+            if form in domain_dict[KEY_ACTIONS]:
+                domain_dict[KEY_ACTIONS].remove(form)
+
+        for key in [KEY_ENTITIES, KEY_ACTIONS, KEY_E2E_ACTIONS]:
+            combined[key] = self.merge_lists(combined[key], domain_dict[key])
+
+        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
+            combined[key] = self.merge_dicts(combined[key], domain_dict[key], override)
+
+        return self.__class__.from_dict(combined)
 
-            merge_dicts.update(b)
-            return merge_dicts
+    def merge_domain_dicts(
+        self, domain1: Dict, domain2: Dict, override: bool = False
+    ) -> Dict[Text, Any]:
+        """Merges this domain dict with another one, combining their attributes.
 
-        def merge_lists(list1: List[Any], list2: List[Any]) -> List[Any]:
-            return sorted(list(set(list1 + list2)))
+        This is used when multiple domain yml files are configured in a single
+        directory. Unlike the merge method above, which merges Domain objects by
```

Comment:
Is there a reason we can't use this implementation for both?

Reply:
@joejuzl So this PR is to port the changes from `2.8.x` into `main`, which we decided to implement in this way as we were planning to [rethink how we handle the domain](https://github.com/RasaHQ/rasa/issues/10510). 

But as we wanted to ensure that `main` also handled multiple domain files, we thought it would be better in the short-term to implement this solution...